### PR TITLE
Sort nodes to group residues

### PR DESCRIPTION
**Pull request by pckroon**
Adds a processor to assign consecutive atomids to molecules, sorted by chain and resid. That way residues will be grouped together in the resulting PDB and ITP file(s).

@fgrunewald Is this more or less as you intended? If yes I'll add some tests.

### CODE REVIEW

File: vermouth/processors/sort_nodes.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+"""
+
+from .processor import Processor
+from ..log_helpers import StyleAdapter, get_logger
+LOGGER = StyleAdapter(get_logger(__name__))
+
+
+class SortNodes(Processor):
+    def __init__(self, groupby=('chain', 'resid', 'insertion_code', 'atomid'),
+                 target_attr='atomid'):
```
Comment (fgrunewald) [inline vermouth/processors/sort_nodes.py:28]: Docstring?

**Review by fgrunewald — CHANGES_REQUESTED** [review]
If I understand correctly it sorts by four factors and the target attribute is the one that is overwritten? This Processor could use some commenting in that regard. Otherwise looks good to me. I would, however, suggest to at least in the doc-string define types that are eligible for group by. It must be a sort able node attribute right like a graph would not work?

**Comment by pckroon** [comment]
It was my plan to write proper docstrings and tests, but I wanted to make sure this is roughly the desired behaviour. I'll push the commits hopefully soon.

**Comment by pckroon** [comment]
`E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/d/dssp/dssp_3.0.0-3build1_amd64.deb  Could not connect to azure.archive.ubuntu.com:80 (20.106.104.242), connection timed out` *sigh*

**Comment by pckroon** [comment]
Tests are working again.

**Review by fgrunewald — APPROVED** [review]
Looks good to me.

**Review by fgrunewald — APPROVED** [review]

**Comment by fgrunewald** [comment]
@pckroon feel free to merge

File: vermouth/processors/sort_molecule_atoms.py [hunk 2]
```diff
@@ -16,41 +16,56 @@
...
+        If not ``None``, new indices will be assigned to this node attribute,
+        starting with 1. It is a good idea to make sure this attribute is also
+        listed in :attr:`sortby_attrs` so that the sorting is stable.
     """
+    def __init__(self, sortby_attrs=('chain', 'resid', 'resname', 'insertion_code', 'atomid'),
+                 target_attr=None):
+        self.sortby_attrs = sortby_attrs
+        self.target_attr = target_attr
+        if self.target_attr is not None and  self.target_attr not in self.sortby_attrs:
+            LOGGER.warning("{} is not in {}. Atom sorting may be unstable.",
```
[the remark this answers is not in the document]
Comment (pckroon) [reply vermouth/processors/sort_molecule_atoms.py:55]: Unstable has a specific meaning in the context of sorting. Consider the following:
`[B1, A2, C1, A1, C3]`. We're going to sort these by letter (numbers are to distinguish the objects with the same letter). A stable sorting will produce `[A2, A1, B1, C1, C3]`, while an unstable sorting *may* swap A2 and A1 and C1 and C3. Stable sorting means that objects which compare as equal remain in the same relative order.

**Pull request merged**